### DISCUSSION TITLE
Use _ID as primary key?

### DISCUSSION THREAD

**Discussion by dgllghr** [kuzudb/kuzu#4444] [Q&A] [answered]
Is there a way to use the autogenerated _ID as the primary key for entities?

**Comment by andyfengHKU** [comment]
Hi @dgllghr,

It's not recommended to use `_ID` as the primary key as it's meant for internal usage and should be read only. May I know what's the main reason that you want to use it as the primary key?

**Reply by semihsalihoglu-uw** [reply]
Also, you may consider using a property of type [SERIAL](https://docs.kuzudb.com/cypher/data-types/#serial) as your primary key.

**Reply by dgllghr** [reply]
Thanks for the quick reply. I have a client application that loads a subgraph from a different graph database for local, user driven analysis using Kuzu. The graph database that the data comes from does not have a primary key, so with Kuzu I have to create a synthetic primary key property for the data, which creates some confusion because that property is not present in the larger source graph. If I could skip the primary key there would be no need for the extra property.

**Reply by semihsalihoglu-uw** [reply]
Interesting. What identifies the nodes in the original/source database? I assume something (maybe a system-level property) must be identifying them.

**Reply by andyfengHKU** [reply]
I believe primary key is necessary because edges need to be defined over unique identifier between nodes. And these unique identifier (which should come from outside of the system) usually serves as the primary key.

**Reply by dgllghr** [reply]
Some of the labels have a unique constraint applied to them, but most are just identified by their system generated ID.

Something else that would work here is the ability to have property level permissions or some kind of VIEW-like functionality (I don't know how that would even work in a graph database though) so I could hide the property. I know this is probably a much bigger thing to consider but I just want to get your thoughts on it. 

BTW, as someone who comes from the relational world, Kuzu makes a lot more sense to me and I have really been enjoying using it.

**Reply by semihsalihoglu-uw** [reply, marked as answer]
I see. So in short, we do need a primary key to be defined somehow to identify the nodes. We might consider having a shortcut that uses a default PK for nodes, i.e., if PK is not defined in the node table schema definition, we add a default SERIAL "ID" or "{tablename}_ID" column, but it would still be visible to you. So not sure what to recommend for now. We would need a longer discussion if having the internal _ID as a default PK if PK is not defined but having PK for node tables is a core assumption in the system, so I'm not very optimistic it would be easy to drop this assumption. We would need to evaluate its feasibility.

For Kuzu looking relational: Great to hear! We are quite open and proud in saying that Kuzu (and frankly almost all other DBMSs) are relational at their cores, even if the data model and query languages have graph-specific features.

**Reply by dgllghr** [reply]
Makes sense. Thanks for the info!
